unit-tests/utilities/time/test-waiting-on.cpp

```diff
@@ -0,0 +1,86 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2021 Intel Corporation. All Rights Reserved.
+
+#include <easylogging++.h>
+#ifdef BUILD_SHARED_LIBS
+// With static linkage, ELPP is initialized by librealsense, so doing it here will
+// create errors. When we're using the shared .so/.dll, the two are separate and we have
+// to initialize ours if we want to use the APIs!
+INITIALIZE_EASYLOGGINGPP
+#endif
+
+#include <unit-tests/catch.h>
+#include <common/utilities/time/waiting-on.h>
+
+using utilities::time::waiting_on;
+
+bool invoke( size_t delay_in_thread, size_t timeout )
+{
+    waiting_on< bool > invoked( false );
+    auto lambda = [delay_in_thread, invoked = invoked.in_thread()]() {
+        //std::cout << "In thread" << std::endl;
+        std::this_thread::sleep_for( std::chrono::seconds( delay_in_thread ));
+        //std::cout << "Signalling" << std::endl;
+        invoked.signal( true );
+    };
+    //std::cout << "Starting thread" << std::endl;
+    std::thread( lambda ).detach();
+    //std::cout << "Waiting" << std::endl;
+    invoked.wait_until( std::chrono::seconds( timeout ), [&]() {
+        return invoked;
+        } );
+    //std::cout << "After wait" << std::endl;
+    return invoked;
+}
+
+TEST_CASE( "Basic wait" )
+{
+    REQUIRE( invoke( 1 /* seconds in thread */, 10 /* timeout */ ) );
+}
+
+TEST_CASE( "Timeout" )
+{
+    REQUIRE( ! invoke( 10 /* seconds in thread */, 1 /* timeout */ ));
+}
+
+TEST_CASE( "Struct usage" )
+{
+    struct value_t
+    {
+        double d;
+        std::atomic_int i{ 0 };
+    };
+    waiting_on< value_t > output;
+    output->d = 2.;
+
+    std::thread( [output_ = output.in_thread()]() {
+        auto p_output = output_.still_alive();
+        auto& output = *p_output;
+        while( output->i < 30 )
+        {
+            std::this_thread::sleep_for( std::chrono::milliseconds( 50 ) );
+            ++output->i;
+            output.signal();
+        }
+    } ).detach();
+
+    // Within a second, i should reach ~20, but we'll ask it top stop when it reaches 10
+    output.wait_until( std::chrono::seconds( 1 ), [&]() {
+        return output->i == 10;
+        } );
+
+    auto i1 = output->i.load();
+    CHECK( i1 >= 10 ); // the thread is still running!
+    CHECK( i1 < 16 );
```

Comment:
how can you be sure i1 < 16 if the thread is still running?

Reply:
If I expect it to notify at 10, 16 would be 6*50=300ms later.
It's possible for it to fail, but very unlikely: it would mean the thread continued but this thread was paused for that amount of time...